orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,486 @@
+	currIdx, _ := bf.getIdxForEndpoint(ec.Endpoint)
+	// TODO: handle above error case
+	endpointsToProbe := []EndpointCriteria{}
+	for i := int64(currIdx + 1); i <= (bf.MaxByzantineNodes + int64(currIdx)); i++ {
+		// add the next (bf.MaxByzantineNodes + 1)
+		endpointsToProbe = append(endpointsToProbe, bf.Endpoints[i%(int64(len(bf.Endpoints)))])
+	}
+
+	// Create an attestation source
+	config := bf.Config
+	config.Endpoints = endpointsToProbe
+	attestation_puller := bf.AttestationSourceFactory(config)
+	defer attestation_puller.Close()
+	attestations := attestation_puller.PullAttestations(seq)
+	return bf.ConfirmByzantineBehavior(attestations)
+}
+
+func (bf *BlockFetcher) isBlockWitheld(seq uint64) bool {
+	// this should keep the current endpoint as reference and try to detect if block witholding
+	// takes place by probing (MaxByzantineNodes + 1) endpoints
+	bf.Logger.Info("Checking for Block withholding")
+	if bf.isEndpointByzantine(bf.currentEndpoint, seq) {
+		bf.suspects.insert(bf.currentEndpoint.Endpoint)
+		return true
+	}
+	return false
+}
+
+func (bf *BlockFetcher) isEndpointSuspected() bool {
+	return bf.suspects.has(bf.currentEndpoint.Endpoint)
+}
+
+func (bf *BlockFetcher) PullBlock(seq uint64) *common.Block {
+	retriesLeft := bf.MaxPullBlockRetries
+
+	// TODO: Should we persist the suspect set across PullBlock ?
+	// For now let us keep it unique for every PullBlock request
+	bf.suspects = suspectSet{}
+
+	for {
+
+		if retriesLeft == 0 && bf.MaxPullBlockRetries > 0 {
+			bf.Logger.Errorf("Failed pulling block [%d]: retry count exhausted(%d)", seq, bf.MaxPullBlockRetries)
+			return nil
+		}
+
+		ec := bf.getSomeEndpoint()
+		start_time := bf.TimeNow()
+		block := bf.pullBlockFromSource(ec, seq)
+
+		if block != nil {
+			if bf.isShuffleTime() {
+				// seems redundant
+				bf.shuffleEndpoint()
+			}
+			return block
+		}
+
+		// if we don't get any block
+		if start_time.Add(time.Duration((int64(bf.Config.FetchTimeout) * bf.ShuffleTimeoutThrehold / 100))).After(time.Now()) {
+			// endpoint might be unavailable so shuffle
+			bf.Logger.Info("couldn't pull block within timeout. endpoint unavailable")
+			bf.shuffleEndpoint()
+			continue
+		}
+
+		// check for blockwitholding and shuffle
+		bf.Logger.Info("Not able to pull block. Check for BFT behavior")
+		if bf.isBlockWitheld(seq) {
+			bf.Logger.Info("BFT behavior detected. Shuffle Endpoint")
+			bf.shuffleEndpoint()
+			continue
+		}
+		// TODO: Here, should we shuffle the endpoint or try from the same endpoint
+
+		retriesLeft--
+	}
+}
+
+func NewBlockFetcher(puller BlockSource) *BlockFetcher {
+	return &BlockFetcher{
+		Endpoints: []EndpointCriteria{},
+		suspects:  suspectSet{},
+	}
+}
+
+func BlockFetcherFromConfigBlock(conf PullerConfig, block *common.Block, verifierRetriever VerifierRetriever, bccsp bccsp.BCCSP) (*BlockFetcher, error) {
+	puller, err := BlockPullerFromConfigBlock(conf, block, verifierRetriever, bccsp)
```

Comment:
if we need only the endpoints from the config block, can we use the utility function directly `EndpointconfigFromConfigBlock`.

Reply:
It is only a placeholder. constructor isn't defined yet.